### PR TITLE
scripts: template json safely

### LINKED ISSUE #853: Safe templating of json in bash scripts
There are a few places in our scripts where we template JSON directly in bash e.g., https://github.com/kubernetes/release/blob/c32c094ff025d4d61fc27e575b489a174090c630/anago#L1131-L1138

As @hoegaarden [mentions](https://github.com/kubernetes/release/pull/849#discussion_r315283633), it'd be nicer/less scary to handle this in a safer way via jq, like this: https://github.com/kubernetes/release/blob/c32c094ff025d4d61fc27e575b489a174090c630/lib/gitlib.sh#L341-L352

This is another good one for a Release Manager Associate to take up.

/help
cc: @kubernetes/release-engineering 
/area release-eng
/priority important-longterm
/kind cleanup

### PR DESCRIPTION
**Pull request by hasheddan**
Updates [anago](https://github.com/kubernetes/release/blob/master/anago) to template json with `jq`.

Fixes #853 

/cc @justaugustus

### CODE REVIEW

File: anago [hunk 1]
```diff
@@ -1133,15 +1133,26 @@ update_github_release () {
...
-  release_id=$($GHCURL $K8S_GITHUB_API/releases$id_suffix --data \
-   '{
-    "tag_name": "'$RELEASE_VERSION_PRIME'",
-    "target_commitish": "'$RELEASE_BRANCH'",
-    "name": "'$RELEASE_VERSION_PRIME'",
-    "body": "See [kubernetes-announce@](https://groups.google.com/forum/#!forum/kubernetes-announce) and ['$CHANGELOG_FILE']('$changelog_url'#'${RELEASE_VERSION_PRIME//\./}') for details.\n\nSHA256 for `kubernetes.tar.gz`: `'${sha256_hash}'`\n\nSHA512 for `kubernetes.tar.gz`: `'${sha512_hash}'`\n\nAdditional binary downloads are linked in the ['$CHANGELOG_FILE']('$changelog_url'#downloads-for-'${RELEASE_VERSION_PRIME//\./}').",
-    "draft": '$draft',
-    "prerelease": '$prerelease'
-    }' |jq -r '.id')
+  query_tmpl='{
```
Comment (hoegaarden) [inline anago]: nit: please make variables you introduce `local`, at the beginning of the func (ideally, IMHO) or when first used.

File: anago [hunk 2]
```diff
@@ -1133,15 +1133,26 @@ update_github_release () {
...
+    "tag_name" : $tag,
+    "target_commitish": $target,
+    "name": $name,
+    "body": $body,
+    "draft": $draft,
+    "prerelease": $prerelease
+  }'
+  body='See [kubernetes-announce@](https://groups.google.com/forum/#!forum/kubernetes-announce) and ['${CHANGELOG_FILE}']('${changelog_url}'#'${RELEASE_VERSION_PRIME//\./}') for details.\n\nSHA256 for `kubernetes.tar.gz`: `'${sha256_hash}'`\n\nSHA512 for `kubernetes.tar.gz`: `'${sha512_hash}'`\n\nAdditional binary downloads are linked in the ['${CHANGELOG_FILE}']('${changelog_url}'#downloads-for-'${RELEASE_VERSION_PRIME//\./}').'
```
Comment (hoegaarden) [inline anago]:
```suggestion
  local body_fmt
  body_fmt='See [kubernetes-announce@](https://groups.google.com/forum/#!forum/kubernetes-announce) and [%s](%s) for details.

SHA256 for `kubernetes.tar.gz`: `%s`
SHA512 for `kubernetes.tar.gz`: `%s`

Additional binary downloads are linked in the [%s](%s).'

body="$(
  printf "$body_fmt" \
    "$CHANGELOG_FILE" "${changelog_url}#${RELEASE_VERSION_PRIME//\./}" \
    "$sha256_hash" "$sha512_hash" \
    "$CHANGELOG_FILE" "${changelog_url}#downloads-for-${RELEASE_VERSION_PRIME//\./}"
)"
```

We could also to something like this, to make that string a little easier to read/understand. What do you think about that?

Comment (hasheddan) [reply anago]: This looks great! Will implement :+1:

**Comment by hasheddan** [comment]
@hoegaarden thanks for your review! I have implemented the suggested changes. In regards to the testing, I heard @justaugustus talk a bit about moving some of these scripts to a more formal tool, likely written in Go. It looks like he has started this effort with #869. I would love to jump in on this effort and potentially take a leadership role in the effort to build a release tool that accomplishes what the collection of scrips currently does. That being said, if it is more desirable just to test these existing scripts, I am happy to do that as well :)

File: anago [hunk 3]
```diff
@@ -1133,15 +1137,37 @@ update_github_release () {
...
+      "$sha256_hash" "$sha512_hash" \
+      "$CHANGELOG_FILE" "${changelog_url}#downloads-for-${RELEASE_VERSION_PRIME//\./}"
+  )"
+  query_body="$( jq \
+    --arg tag "$RELEASE_VERSION_PRIME" \
+    --arg target "$RELEASE_BRANCH" \
+    --arg name "$RELEASE_VERSION_PRIME" \
+    --arg body "$body" \
+    --arg draft "$draft" \
+    --arg prerelease "$prerelease" \
```
Comment (hoegaarden) [inline anago]: `draft` & `prerelease` [should be booleans](https://developer.github.com/v3/repos/releases/#create-a-release), therefore we should:
- use `--argjson` for those
- maybe (didn't trace back, where this variables come from) have a function `common::validate_bool` or such to validate if the strings we use here are valid JSON bools?

File: anago [hunk 4]
```diff
@@ -1133,15 +1137,37 @@ update_github_release () {
...
+    "body": $body,
+    "draft": $draft,
+    "prerelease": $prerelease
+  }'
+  body_fmt='See [kubernetes-announce@](https://groups.google.com/forum/#!forum/kubernetes-announce) and [%s](%s) for details.
+  
+    SHA256 for `kubernetes.tar.gz`: `%s` 
+    SHA512 for `kubernetes.tar.gz`: `%s`
+  
+    Additional binary downloads are linked in the [%s](%s).'
```
Comment (hoegaarden) [inline anago]: nit: I am not sure, if the leading spaces here will be ignored anyway (and the trailing space in line 1150), but I think I'd prefer removing them.

**Comment by hoegaarden** [comment]
@hasheddan 

> [...] That being said, if it is more desirable just to test these existing scripts, I am happy to do that as well :)

First off: I am definitely not asking you to backfill tests for the entirety of `anago` :)

> [...] In regards to the testing, I heard @justaugustus talk a bit about moving some of these scripts to a more formal tool

Yep, I am aware that there are some efforts going on to replace `anago` and others with existing and new tools. However, I guess that will still take some time.
So I'd love to get some tests in place for stuff that we change, so that I have some more confidence that the scripts won't break for some other member of the (patch-)release-team after I merge something. Therefore I was suggesting to add some tests around the generation of the payload; we could capture the current payload and check that we generate the same one after our refactor. That would give me way more confidence when reviewing.
But still, I won't block on that. :)

File: anago (hunk 3 above)
Comment (hasheddan) [reply anago]: Looks like we set both of these variables explicitly in the preceding lines so I believe they should be safe here

Comment (hasheddan) [reply anago]: Agreed to using `--argjson` for sure though!

**Pull request merged**